build_zng.rs

```diff
@@ -1,58 +1,691 @@
-use std::env;
+use std::{
+    collections::{BTreeMap, BTreeSet},
+    env, fs,
+    hash::{Hash as _, Hasher as _},
+    io::Write as _,
+    path::{Path, PathBuf},
+};
+
+fn append(cfg: &mut cc::Build, root: Option<&str>, files: impl IntoIterator<Item = &'static str>) {
+    let root = root.unwrap_or("");
+    cfg.files(
+        files
+            .into_iter()
+            .map(|fname| format!("src/zlib-ng/{root}{fname}.c")),
+    );
+}
+
+/// Replicate the behavior of cmake/make/configure of stripping out the
+/// @ZLIB_SYMBOL_PREFIX@ since we don't want or need it
+fn strip_symbol_prefix(input: &Path, output: &Path, mut on_line: impl FnMut(&str)) {
```

Comment:
I suggest using trait object for faster compilation:

```suggestion
fn strip_symbol_prefix(input: &Path, output: &Path, on_line: & dyn mut FnMut(&str)) {
```